services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -219,6 +219,11 @@ public Response verifyUserCode(@QueryParam("user_code") String userCode) {
                 return createVerificationPage(Messages.OAUTH2_DEVICE_EXPIRED_USER_CODE);
             }
 
+            if (deviceCode.isDenied()) {
+                event.error(Errors.INVALID_OAUTH2_USER_CODE);
+                return createVerificationPage(Messages.OAUTH2_DEVICE_CONSENT_DENIED);
```

Comment:
Or use invalid user code error? I wasn't sure.

Reply:
At this stage, we should probably give a user code error, similar to what you did to the pending state.

Reply:
+1 for (2) and (3)

Reply:
Done (options 2 and 3). Thanks for the input!